### ISSUE TITLE
Feature request: static bit_length() method on types

### ISSUE THREAD

**Issue by Shnatsel** [phil-opp/rust-bit-field#7]
I'm using the remainder from a big integer to get position in a bit field, like so:

`position_in_block = input % RemainderFromInputGetsYouPositionInBitField;`

The bit field size is known at compile time, but I want to be able to easily tune it, so I've tried the following (type conversion omitted):

`position_in_block = input % PackedBlock::from(0).bit_length();`

This works (and I imagine it's very useful in generics), but incurs significant runtime overhead compared to a hardcoded value (this is in a hot codepath and total execution time went up 20% because of it). I've also tried putting the same thing in a constant, but the compiler does not allow that either: "calls in constants are limited to constant functions, struct and enum constructors".

Adding static bit_length() methods on primitive types would be very useful for such use cases. This is the last thing holding me back from completely abstracting the code from the size of the bitfield.

**Comment by phil-opp** [comment]
Sorry for the late reply!

So you're proposing to make `bit_length` a `const fn`? I would love to do that, but unfortunately it's not possible to make trait functions `const`.

It's a bit strange that the compiler didn't inline this in your case. Did you compile with optimizations (i.e. in release mode)?

**Comment by Shnatsel** [comment]
Yeah, I'm sure I compiled it in release mode. I had a somewhat weird LLVM target though (asm.js), I'm not sure if that could affect inlining or not.

**Comment by phil-opp** [comment]
Hmm, I'm not sure either. Could be that the LLVM optimizer isn't that mature for asm.js yet… Maybe ask on the Rust IRC channel?

**Comment by phil-opp** [comment]
Is this still a problem @Shnatsel?

With #10, the `bit_length` method is now independent of `&self`, i.e. `u32::bit_length()` instead of `0u32.bit_length()`. I also opened #13 to turn it into an associated `BIT_LENGTH` constant, but we can't merge it until the `const_size_of` feature becomes stable.

**Comment by phil-opp** [comment]
Closing in favor of #13.
